### ISSUE TITLE
Problem in Soro tree counts?

### ISSUE THREAD

**Issue by florianhartig** [COST-FP1304-PROFOUND/ProfoundData#18]
Reported from Mats: One particularity I found in the DB itself is that in the TREES table for Soro in 2009 there is double the amount of trees than in the year before and every record_id is duplicated. At the moment I can not figure out why - should I open an issue even though that problem is not related to the package itself?

**Comment by florianhartig** [comment]
Comment Ramiro: 

> Concerning this, a bit of the error checking is done by the code in the test folder. There are test for the package ( and test for the DB (/PROFOUND database/testDB), and actually duplicates are part of the test(which seem to now work?). An overview of the data, for visual inspection, can be done with /data-raw/PROFOUNDdatabase.Rmd, which creates the overview vignette. An overview of all sites is provided by "reportDB" (pkg function). I attached the oouput of reportDB to this email (PROFOUNDsites.html), and I see in the file the doubled tree amount of Soro in 2009.
> I do not find the duplicates in "Tree_Data.RData" file and I do not see how the duplicates could have been created (see "Tree_to_DB.R). I am clueless here. Anyway, the duplicated records should be deleted.

**Comment by mmahnke** [comment]
Hi guys. I think I found the problem.

The TREE_master table in the DB has no duplicates, TREE and TREE_[site_id] (e.g. site_id=21 which is Soro) views have duplicates. If I understand correctly, these views are used when extracting data via getData(), that is why there are duplicated records.

> TREES <- RSQLite::dbGetQuery(con, "SELECT * FROM TREE_master")
> length(which(duplicated(TREES)))
[1] 0

>TREES <- RSQLite::dbGetQuery(con, "SELECT * FROM TREE")
> length(which(duplicated(TREES))) 
[1] 20650 

> TREES <- RSQLite::dbGetQuery(con, "SELECT * FROM TREE_21") 
> length(which(duplicated(TREES))) 
[1] 288

The duplicates are introduced in "Create_TREE_Views.R" when PLOTSIZE_master and TREE_master are joined to create the views because there are multiple entries of plotsize in PLOTSIZE_master for the same site and year with different record_ids.
>dbGetQuery([...] AND  TREE_master.year = PLOTSIZE_master.year)

Multiple observations of plotsize for the same year and plot come from the processed input data in "~/ownCloud/PROFOUND_Data/Processed/RData/PlotSize_Data.RData" that is used in "Create_TREE_Views.R". BTW, there are also duplicates for Hyytiala and KROOF, I guess those TREE views are also not correct.

In summary, I suggest to edit the "PlotSize_Data.RData" in the owncloud so that for each site-year combination there is only one observation and then rerun the create-PROFOUND-database-scripts to produce a new sqlite DB that can then be distributed.

**Comment by florianhartig** [comment]
Sounds convincing to me, Ramiro?

Mats, can you just do this?

**Comment by mmahnke** [comment]
Yes, I'll do that.

**Comment by christopherreyer** [comment]
sounds good to me, Mats, once you have a new version of the DB, we then also need to update the doi, please remind me...

**Comment by mmahnke** [comment]
I cannot quite finish producing the SQL DB. The _Create_DB.R_ throws warnings/errors that I cannot resolve on my own at the moment. @Rasilgon can you help out?

What I have done for now:
1. change _Create_PlotSize_Data.R_ to prevent duplicates: #22 
2. add new version number "0.2.1" in DB versioning in _owncloud/PROFOUND_Data/Processed/Version/version.csv_ and _owncloud/PROFOUND_Data/Processed/RData/Version_Data.RData_
3. updated _owncloud/PROFOUND_Data/Processed/RData/PlotSize_Data.RData_ using the changed _Create_PlotSize_Data.R_

Then trying to create the SQL DB by _Create_DB.R_, I get lots of warnings like:
1. `Closing open result set, pending rows`
2. `In result_fetch(res@ptr, n = n) : SQL statements must be issued with dbExecute() or dbSendStatement() instead of dbGetQuery() or dbSendQuery()`
3. and one single error thrown from _FLUXNET_to_DB.R_ that comes from
`dbWriteTable(db, "FLUXNET_master", df[,columns], append=TRUE, row.names = FALSE) ` (l.177)
indicating that _colnames_ do not match between _FLUXNET_master_ and _df[,columns]_

**Comment by Rasilgon** [comment]
Hi @matsmahnken,

are you running the script _Create_DB,R_ at one time? I normally by one by one the code lines, to supervise that each step works.

Warnings are not errors necessarily. Concerning the three issues you measure, I can tell you that


1. It is displayed when sending queries to the DB. It is been a bit cryptic for me also, but does not mean an error. At least, so far it did not mean that less data was entered or download from the DB.

2. Where exactly is the warning coming from? I cannot find/reproduce it.

3. At some point, in a newer version of dbWriteTable(DBI) they introduced code to check whether the column names of the data and the column names of the TABLE in the DB match. The last time I was working on it, I reckon that it was only what a warning saying that they should meet, but I assume that in newer version of the function it might through an error. So this should be taken care of. The solutions is to rename the column names of FLUXNET data files (physically or virtually) before entering the data in the db, so they match.

**Comment by matsmahnken** [comment]
Hi @Rasilgon,

thanks for the notes! For now I have been running the _Create_DB.R_ code line by line. And I get warning 1 (about closing open results)  and warning 2 (sql issues must be issued...) for most of the lines. And point 3 occurs when sourcing _FLUXNET_to_DB.R_.

1. From now on I will ignore warning 1 since it does not mean error, as you noted.
2. Looking into the scripts I am sourcing I see that the warning 2 comes from running _dbGetQuery_, but I was not able to find out why. Maybe I can also ignore that since it is not an error?! Could it have something to do with my R version or the package version?
3. To solve the problem, can we insert code to rename the colnames of _FLUXNET_raw_ in _PROFOUND database/Processing/FLUXNET_rawedit.R_ [here](https://github.com/matsmahnken/ProfoundData/blob/ae7951c0a030c0f0443f2ae996f6a159f846d2ce/PROFOUND%20database/Processing/FLUXNET_rawedit.R#L71)?

**Comment by florianhartig** [comment]
Hi guys, is this issue still open?

**Comment by matsmahnken** [comment]
Yes, still open. Although this issue could be split in two. The first problem on the duplicated records is solved I guess, the second problem regarding building the new version of the sqlite DB without duplicated tree records is still open for me - I am currently working on it. I hope I can give some updates next week.

But these open DB building problems do actually not relate to the package itself.

**Comment by florianhartig** [comment]
Ok, thanks. Then we should just leave it open. In the end, it's more a DB problem than a package problem, maybe we should create different milestones for that.

**Comment by matsmahnken** [comment]
Finally I successfully produced a good-looking sqlite DB. All my trouble was coming from encoding problems from carrying over the scripts and .RData files to Windows. Running the scripts to produce the DB in a Linux environment solved all issues.
The warnings and FLUXNET_master colname problems disappeared.

**Comment by matsmahnken** [comment]
This issue is solved in the new DB version 0.3
